RuddockWebsite/auth_utils.py

```diff
@@ -0,0 +1,291 @@
+import hashlib
+import passlib.hash
+import string
+from sqlalchemy import text
+from flask import session, g, url_for, flash
+from RuddockWebsite import constants
+from RuddockWebsite import email_utils
+from RuddockWebsite import email_templates
+from RuddockWebsite import misc_utils
+
+class PasswordHashParser:
+  '''
+  Class to manage parsed password hashes.
+
+  Handling of current and legacy hashing algorithms:
+  ==================================================
+
+  We store all hashes in the format:
+
+    $algorithm1|...|algorithmN$rounds1|...|roundsN$salt1|...|saltN$hash
+
+  Where algorithmN is the most current algorithm (think of it as piping the
+  result of one algorithm to the next). To compute the hash, we execute:
+
+    algorithmN(saltN + algorithmN-1(saltN-1 + ... algorithm1(salt1 + password)))
+
+  We do not store hashes from legacy algorithms in the database (the whole
+  point of upgrading algorithms is to improve hash strength). Instead, we use
+  the output from the first hash as the 'password' input to the next hash, so
+  all passwords are hashed with the more secure algorithm. To keep track of the
+  algorithm and parameters (rounds, salt) for each step, we use the format
+  above. For example, if a password is initally hashed with MD5, then later
+  upgraded to SHA1, then later upgraded to PBKDF2_SHA256, the hash string might
+  look like this:
+
+    $md5|sha1|pbkdf2_sha256$||100000$salt1|salt2|salt3$final_hash_here
+
+  Note that 'rounds' may be the empty string if it does not apply to the
+  algorithm used.
+
+  Storing hashes in this format was designed to make it easier to upgrade
+  algorithms in the future (simply take the actual hash as the password,
+  generate a new salt, and then append the new algorithm, number of rounds, and
+  salt to the modular hash formatted string). Upgrading algorithms with this
+  design should be invisible to users.
+  '''
+
+  # Static list of supported hashing algorithms.
+  valid_algorithms = ['md5', 'pbkdf2_sha256']
+
+  def __init__(self, algorithms=[], rounds=[], salts=[], password_hash=None):
```

Comment:
Not sure we should have these defaults here. Presumably there must be at least one algorithm.